net/portmapper/portmapper.go

```diff
@@ -490,13 +495,32 @@ func (c *Client) createMapping() {
 		c.runningCreate = false
 	}()
 
-	if _, err := c.createOrGetMapping(ctx); err == nil && c.onChange != nil {
+	mapping, _, err := c.createOrGetMapping(ctx)
+	if err != nil {
+		if !IsNoMappingError(err) {
+			c.logf("createOrGetMapping: %v", err)
+		}
+		return
+	}
+	c.updates.Publish(Mapping{
+		External:  mapping.External(),
+		Type:      mapping.MappingType(),
+		GoodUntil: mapping.GoodUntil(),
+	})
+	if c.onChange != nil {
 		go c.onChange()
-	} else if err != nil && !IsNoMappingError(err) {
-		c.logf("createOrGetMapping: %v", err)
 	}
 }
 
+// MappingEvent is an event recording the allocation of a
+type Mapping struct {
+	External  netip.AddrPort
+	Type      string
+	GoodUntil time.Time
+
+	// TODO(creachadair): Record whether we reused an existing mapping?
```

Comment:
Hmm, good question. My first thought would be to go somewhat the other way: remove GoodUntil from the event, and have the mapper emit a corresponding MappingLost event when it deems it appropriat. Then the mapper can also handle deduplication (i.e. don't re-emit an event if it's re-upped an existing mapping and nothing's changed).

But, that's all very much TODO-grade thoughts, the sucky bit is plumbing events at all, iterating on event shape afterwards is much more fun :)

Reply:
Yup, agreed.